index.bs

```diff
@@ -3763,10 +3771,678 @@ nothrow>WritableStreamDefaultControllerError ( <var>controller</var>, <var>error
+  1. If Type(_x_) is not Object, return *false*.
+  1. If _x_ does not have a [[transformStreamController]] internal slot, return *false*.
+  1. Return *true*.
+</emu-alg>
+
+<h4 id="transform-stream-error" aoid="TransformStreamError" nothrow>TransformStreamError ( <var>stream</var>,
+<var>e</var> )</h4>
+
+<emu-alg>
+  1. If _stream_.[[readable]].[[state]] is `"readable"`, perform !
+     ReadableStreamDefaultControllerError(_stream_.[[readable]].[[readableStreamController]], _e_).
+  1. Perform ! TransformStreamErrorWritableAndUnblockWrite(_stream_, _e_).
+</emu-alg>
+
+<p class="note">This operation works correctly when one or both sides are already errored. As a result, calling
+algorithms do not need to check stream states when responding to an error condition.</p>
+
+<h4 id="transform-stream-error-writable-and-unblock-write" aoid="TransformStreamErrorWritableAndUnblockWrite"
+nothrow>TransformStreamErrorWritableAndUnblockWrite ( <var>stream</var>, <var>e</var> )</h4>
+
+<emu-alg>
+  1. Perform ! WritableStreamDefaultControllerErrorIfNeeded(_stream_.[[writable]].[[writableStreamController]], _e_).
+  1. If _stream_.[[backpressure]] is *true*, perform ! TransformStreamSetBackpressure(_stream_, *false*).
+</emu-alg>
+
+<p class="note">The {{TransformStreamDefaultSink/write()}} method of {{TransformStreamDefaultSink}} could be waiting for
+the promise stored in the \[[backpressureChangePromise]] slot to resolve. This call to TransformStreamSetBackpressure
+ensures that the promise always resolves.</p>
+
+<h4 id="transform-stream-set-backpressure" aoid="TransformStreamSetBackpressure" nothrow>TransformStreamSetBackpressure
+( <var>stream</var>, <var>backpressure</var> )</h4>
+
+<emu-alg>
+  1. Assert: _stream_.[[backpressure]] is not _backpressure_.
+  1. If _stream_.[[backpressureChangePromise]] is not *undefined*, <a>resolve</a>
+     stream.[[backpressureChangePromise]] with *undefined*.
+  1. Set _stream_.[[backpressureChangePromise]] to <a>a new promise</a>.
+  1. Set _stream_.[[backpressure]] to _backpressure_.
+</emu-alg>
+
+<h3 id="ts-default-controller-class" interface lt="TransformStreamDefaultController">Class
+<code>TransformStreamDefaultController</code></h3>
+
+The {{TransformStreamDefaultController}} class has methods that allow manipulation of the associated {{ReadableStream}}
+and {{WritableStream}}. When constructing a {{TransformStream}}, the <a>transformer</a> is given a corresponding
+{{TransformStreamDefaultController}} instance to manipulate.
+
+<h4 id="ts-default-controller-class-definition">Class Definition</h4>
+
+<div class="non-normative">
+
+<em>This section is non-normative.</em>
+
+If one were to write the {{TransformStreamDefaultController}} class in something close to the syntax of [[!ECMASCRIPT]],
+it would look like
+
+<pre><code class="lang-javascript">
+  class TransformStreamDefaultController {
+    constructor(stream)
+
+    get desiredSize()
+
+    enqueue(chunk)
+    error(reason)
+    terminate()
```

Comment:
I find it interesting that we have `cancel()`, `terminate()`, and `close()`.

Reply:
Yes, they all have different semantics. So based on the principle of "different things should look different", we use different names for them. See #774 for the discussion where we added `terminate()`. The difference between `cancel()` and `close()` is mentioned in the FAQ: https://github.com/whatwg/streams/blob/master/FAQ.md#why-dont-errors-that-occur-while-cancelling-put-the-readable-stream-in-an-error-state

Reply:
That's awesome. Definitely a lesson in there for me. Thanks, @ricea.